src/ripple/rpc/handlers/GetAggregatePrice.cpp

```diff
@@ -0,0 +1,321 @@
+#include <ripple/app/main/Application.h>
+#include <ripple/json/json_value.h>
+#include <ripple/ledger/ReadView.h>
+#include <ripple/protocol/ErrorCodes.h>
+#include <ripple/protocol/jss.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/impl/RPCHelpers.h>
+
+#include <boost/bimap.hpp>
+#include <boost/bimap/multiset_of.hpp>
+
+namespace ripple {
+
+struct CompareDescending
+{
+    bool
+    operator()(std::uint32_t const& a, std::uint32_t const& b) const
+    {
+        return a > b;  // Sort lastUpdateTime in descending order
+    }
+};
+
+using namespace boost::bimaps;
+// sorted descending by lastUpdateTime, ascending by AssetPrice
+using Prices =
+    bimap<multiset_of<std::uint32_t, CompareDescending>, multiset_of<STAmount>>;
+
+/** Calls callback "f" on the ledger-object sle and up to three previous
+ * metadata objects. Returns if the callback returns true.
+ */
+static void
+iteratePriceData(
+    RPC::JsonContext& context,
+    std::shared_ptr<SLE const> const& sle,
+    std::function<bool(STObject const&)>&& f)
+{
+    using Meta = std::shared_ptr<STObject const>;
+    constexpr std::uint8_t maxHistory = 3;
+
+    if (!sle || f(*sle))
+        return;
+
+    uint256 prevTx = sle->getFieldH256(sfPreviousTxnID);
+    std::uint32_t prevSeq = sle->getFieldU32(sfPreviousTxnLgrSeq);
+    std::uint8_t history = 0;
+
+    // return tx meta-data for prevSeq/prevTx if the history
+    // does not exceed the max
+    auto getMeta = [&](std::uint32_t prevSeq, uint256 const& prevTx) -> Meta {
+        if (++history <= maxHistory)
+        {
+            if (auto const ledger =
+                    context.ledgerMaster.getLedgerBySeq(prevSeq))
+                return ledger->txRead(prevTx).second;
+        }
+        return nullptr;
+    };
+
+    while (Meta const& meta = getMeta(prevSeq, prevTx))
+    {
+        for (STObject const& node : meta->getFieldArray(sfAffectedNodes))
+        {
+            if (node.getFieldU16(sfLedgerEntryType) == ltORACLE)
+            {
+                bool const isNew = node.isFieldPresent(sfNewFields);
+                // if a meta is for the new and this is the first
+                // look-up then it's the meta for the tx that
+                // created the current object; i.e. there is no
+                // historical data
+                if (isNew && history == 1)
+                    return;
+
+                auto const& n = isNew ? static_cast<const STObject&>(
+                                            node.peekAtField(sfNewFields))
+                                      : static_cast<const STObject&>(
+                                            node.peekAtField(sfFinalFields));
+
+                // there is no previous seq/tx if new
+                if (f(n) || isNew)
+                    return;
+
+                prevTx = node.getFieldH256(sfPreviousTxnID);
+                prevSeq = node.getFieldU32(sfPreviousTxnLgrSeq);
+                break;
+            }
+        }
+    }
+}
```

Comment:
This looks much more complicated than I think it needs to be. I've rewritten it below. What do you think? Are my added comments correct? More questions follow.

```cpp
static void
iteratePriceData(
    RPC::JsonContext& context,
    SLE const* oracle,
    std::function<bool(STObject const&)>&& f)
{
    using Meta = std::shared_ptr<STObject const>;
    constexpr std::uint8_t maxHistory = 3;

    // `oracle` points to an object that has an `sfPriceDataSeries` field.
    // When this function is called, that is a `PriceOracle` ledger object,
    // but after one iteration of the loop below, it is an `sfNewFields`
    // / `sfFinalFields` object in a `CreatedNode` / `ModifiedNode` object in
    // a transaction's metadata.

    // `chain` points to an object that has `sfPreviousTxnID` and
    // `sfPreviousTxnLgrSeq` fields. When this function is called,
    // that is the `PriceOracle` ledger object pointed to by `oracle`,
    // but after one iteration of the loop below, then it is a `ModifiedNode`
    // / `CreatedNode` object in a transaction's metadata.
    STObject const* chain = oracle;
    std::uint8_t history = 0;
    bool isNew = false;
    while (true) {
        if (!oracle || f(*oracle) || isNew)
            return;

        if (++history > maxHistory)
            return;

        uint256 prevTx = chain->getFieldH256(sfPreviousTxnID);
        std::uint32_t prevSeq = chain->getFieldU32(sfPreviousTxnLgrSeq);

        auto const ledger = context.ledgerMaster.getLedgerBySeq(prevSeq);
        if (!ledger)
            return;

        Meta const& meta = ledger->txRead(prevTx).second;

        for (STObject const& node : meta->getFieldArray(sfAffectedNodes))
        {
            if (node.getFieldU16(sfLedgerEntryType) != ltORACLE)
            {
                continue;
            }

            chain = &node;
            isNew = node.isFieldPresent(sfNewFields);
            // if a meta is for the new and this is the first
            // look-up then it's the meta for the tx that
            // created the current object; i.e. there is no
            // historical data
            if (isNew && history == 1)
                return;

            oracle = isNew
                ? static_cast<const STObject&>(node.peekAtField(sfNewFields))
                : static_cast<const STObject&>(node.peekAtField(sfFinalFields));
            break;
        }
    }
}
```

1. I haven't yet checked where this is called, but I find the behavior around the check `isNew && history == 1` surprising. I think I understand why it is there: when `iteratePriceData` is called, it is passed an `Oracle` ledger object, which is passed directly to the callback, and if the first transaction in its history is the one that created it, then the `CreatedNode` will have the exact same price series as the object that was passed, and we don't want to call the callback on it twice. Is that why? If it is, then isn't that always true of the first transaction in the history, even when it is a `ModifiedNode`? Won't the `sfFinalFields` of the latest transaction have the exact same price series as in the object that was passed? Should the check we do here instead be something like this?

```
if (history == 1) {
    if (isNew) return;
    else continue with the loop but skip the callback on the next iteration;
}
```

2. This is a minor nit. The inner loop does not have the best behavior if the `CreatedNode` / `ModifiedNode` for the `Oracle` is not found. Technically that should never be possible in the data, but it might be worth a sanity check. In that case, we can store the previous value of `chain` in a separate variable, initialized to `nullptr`, and check that the value of `chain` is different from it near the start of every iteration of the outer loop.

Reply:
The comments sound right but there are two issues with your code. One is easy to fix where `STObject` object is assigned to `STLedgerEntry` here 

    oracle = isNew
                ? static_cast<const STObject*>(&node.peekAtField(sfNewFields))
                : static_cast<const STObject*>(&node.peekAtField(sfFinalFields));

Another issue is more difficult and is the reason I structured this function the way I did. The problem is that `oracle` on this line

    if (!oracle || f(*oracle) || isNew)

 is undefined on the second iteration because `meta` goes out of scope and so do all members that `meta` owns. I have not had much time today because of the meeting, etc. to think about simplifying this function. I'll give it more thought tomorrow.

1. You are right about the intention but not the other assumption. If there is only one object on the ledger, then the metadata points to the same object because the previous fields of the ledger object that is passed in are pointing to this same object. If there are two objects then the previous fields of the ledger object point to the previous object. If we use your suggested check then we'll skip this second object.
2. You are right. I added for now an extra field `lastPrevSeq` to check for this condition.

Reply:
We can do something like this:
   
    Meta meta = nullptr;
    while (true) {
        if (!oracle || f(*oracle) || isNew)
            return;
        ...
        meta = ledger->txRead(prevTx).second;
        ...
    }

But then the `meta` shared pointer is going to be copied.